Sources/NIOHTTP1/HTTPDecoder.swift

```diff
@@ -358,91 +376,120 @@ public class HTTPDecoder<HTTPMessageT>: ByteToMessageDecoder, AnyHTTPDecoder {
 
-            state.baseAddress = nil
+                    state.baseAddress = nil
 
-            let httpError = parser.http_errno
-            if httpError != 0 {
-                self.state.currentError = HTTPParserError.httpError(fromCHTTPParserErrno: http_errno(rawValue: httpError))!
-                throw self.state.currentError!
+                    if let error = state.currentError {
+                        throw error
+                    }
+
+                    let httpError = parser.http_errno
+                    if httpError != 0 {
+                        self.state.currentError = HTTPParserError.httpError(fromCHTTPParserErrno: http_errno(rawValue: httpError))!
+                        throw self.state.currentError!
+                    }
+                    return result
+                }
+
+                guard self.cumulationBuffer != nil else {
+                    // Guard against the case of closing the channel while still in the decode loop. In this case the cumulationBuffer will be nil.
+                    return
+                }
+
+                /// http_parser_execute(...) should always consume the whole buffer.
+                assert(result == bufferSlice.readableBytes)
+
+                // Update readerIndex of the cumulationBuffer itself as we will refetch it in the next loop run if needed.
+                self.cumulationBuffer!.moveReaderIndex(forwardBy: result)
+
+                if bufferSlice.readableBytes == self.cumulationBuffer!.readableBytes + result {
+                    // There was no change to the readable bytes of the cumulationBuffer by a re-entrant call of channelRead. Its safe to break the loop.
+                    break
+                }
+            } while true
+
+            switch state.dataAwaitingState {
+            case .body, .messageBegin:
+                // Ensure there is no risk to invalidate any indices that we depend on.
+                assert(state.currentNameIndex == nil)
+                assert(state.currentHeaders.isEmpty)
+                assert(state.slice == nil)
+
+                // Only in the body and messageBegin states its safe to discard read bytes as otherwise we may risk to invalidate the stored indices for the
+                // Headers and slice.
+                if self.shouldReclaimBytes(buffer: self.cumulationBuffer!) {
+                    self.cumulationBuffer!.discardReadBytes()
+                }
+            default:
+                // TODO: We may be able to also discard some read bytes in other states if we are very careful about it.
+                //       What we would need to do is basically keep track of the start of a message and discard everything that we already forwarded through the pipeline.
+                break
             }
-            return result
+        } catch {
+            self.cumulationBuffer = nil
+            ctx.fireErrorCaught(error)
+            ctx.close(promise: nil)
         }
+    }
```

Comment:
This method's 80 lines long, can we factor it into some smaller sub-functions?

Reply:
done